src/error.rs

```diff
@@ -1,84 +1,515 @@
+    /// or that no format could be determined, or that it did not match format specific
+    /// requirements set by the caller.
+    Decoding(DecodingError),
 
-    /// The Image's dimensions are either too small or too large
-    DimensionError,
+    /// An error was encountered while encoding.
+    ///
+    /// The input image can not be encoded with the chosen format, for example because the
+    /// specification has no representation for its color space or because a necessary conversion
+    /// is ambiguous. In some cases it might also happen that the dimensions can not be used with
+    /// the format.
+    Encoding(EncodingError),
 
-    /// The Decoder does not support this image format
-    UnsupportedError(String),
+    /// An error was encountered in inputs arguments.
+    ///
+    /// This is a catch-all case for strictly internal operations such as scaling, conversions,
+    /// etc. that involve no external format specifications.
+    Parameter(ParameterError),
 
-    /// The Decoder does not support this color type
-    UnsupportedColor(ExtendedColorType),
+    /// Completing the operation would have required more resources than allowed.
+    ///
+    /// Errors of this type are limits set by the user or environment, *not* inherent in a specific
+    /// format or operation that was executed.
+    Limits(LimitError),
 
-    /// Not enough data was provided to the Decoder
-    /// to decode the image
-    NotEnoughData,
+    /// An operation can not be completed by the chosen abstraction.
+    ///
+    /// This means that it might be possible for the operation to succeed in general but
+    /// * it requires a disabled feature,
+    /// * the implementation does not yet exist, or
+    /// * no abstraction for a lower level could be found.
+    Unsupported(UnsupportedError),
 
-    /// An I/O Error occurred while decoding the image
+    /// An error occurred while interacting with the environment.
     IoError(io::Error),
+}
 
-    /// The end of the image has been reached
-    ImageEnd,
+/// The implementation for an operation was not provided.
+///
+/// See the variant [`Unsupported`] for more documentation.
+///
+/// [`Unsupported`]: struct.ImageError.html#variant.Unsupported
+#[derive(Debug)]
+pub struct UnsupportedError {
+    format: ImageFormatHint,
+    kind: UnsupportedErrorKind,
+}
+
+/// Details what feature is not supported.
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum UnsupportedErrorKind {
+    /// The required color type can not be handled.
+    Color(ExtendedColorType),
+    /// An image format is not supported.
+    Format(ImageFormatHint),
+    /// Some feature specified by string.
+    /// This is discouraged and is likely to get deprecated (but not removed).
+    GenericFeature(String),
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+#[derive(Debug)]
+pub struct EncodingError {
+    format: ImageFormatHint,
+    message: String,
```

Comment:
Also by @Nemo157

> one minor optimization is to replace String variants with Box<str> since you don't need to mutate them, that saves a third of the data size in those variants

Reply:
This would be used in the internal representations only, not those `*Kind` enums exposed.